TestLauncher cannot run the same test in different test tasks if configuration cache is enabled

### Current Behavior

Running a test method twice in different test tasks result in the following exception:

```
Reusing configuration cache.
FAILURE: Build failed with an exception.
* What went wrong:
Task with name 'forkingIntegTest' not found in project ':language-java'.
* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Get more help at https://help.gradle.org.
BUILD FAILED in 353ms
Publishing build scan...
https://ge.gradle.org/s/nnkvf236omh5u
Configuration cache entry reused.
```

### Expected Behavior

No Exception.
The configuration cache entry should not be reused if a different test task is specified via the TestLauncher.

### Context (optional)

IntelliJ IDEA has support for invoking Gradle tests via the `TestLauncher` API that will be eventually turned on by default. I assigned this issue to the next RC1 milestone as fixing this issue later might delay the adoption further.

### Steps to Reproduce

1. Open IntelliJ IDEA 2023.2 and import the gradle/gradle repository
2. In the menu, click Help > Edit Custom Properties... and add `gradle.testLauncherAPI.enabled=true`
3. Restart IDEA
4. Go to `JavaIncrementalCompilationWithCachingIntegrationTest` and run the `incremental state is restored from the build cache` with the green play button. Select the `embeddedIntegTest` task and observe the passing test.
5. Edit the freshly created run configuration and replace `embeddedIntegTest` with `forkingIntegTest`.
6. Execute the modified run configuration and observe the build failure.

### Gradle version

8.4-20230719142735+0000

### Build scan URL (optional)

https://ge.gradle.org/s/nnkvf236omh5u

### Your Environment (optional)

_No response_

--- a/subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/action/TestExecutionRequestAction.java
+++ b/subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/action/TestExecutionRequestAction.java
@@ -16,9 +16,12 @@
 
 package org.gradle.tooling.internal.provider.action;
 
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import org.gradle.api.internal.StartParameterInternal;
+import org.gradle.internal.DefaultTaskExecutionRequest;
+import org.gradle.internal.RunDefaultTasksExecutionRequest;
 import org.gradle.internal.build.event.BuildEventSubscriptions;
 import org.gradle.tooling.events.test.internal.DefaultDebugOptions;
 import org.gradle.tooling.internal.protocol.events.InternalTestDescriptor;
@@ -34,6 +37,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 public class TestExecutionRequestAction extends SubscribableBuildAction {
     private final StartParameterInternal startParameter;
@@ -69,14 +73,42 @@ public class TestExecutionRequestAction extends SubscribableBuildAction {
     // TestExecutionRequestAction
     public static TestExecutionRequestAction create(BuildEventSubscriptions clientSubscriptions, StartParameterInternal startParameter, ProviderInternalTestExecutionRequest testExecutionRequest) {
         ImmutableSet<String> classNames = ImmutableSet.copyOf(testExecutionRequest.getTestClassNames());
-        return new TestExecutionRequestAction(clientSubscriptions, startParameter,
+        List<InternalTaskSpec> taskSpecs = testExecutionRequest.getTaskSpecs(Collections.emptyList());
+        boolean runDefaultTasks = testExecutionRequest.isRunDefaultTasks(false);
+
+        return new TestExecutionRequestAction(
+            clientSubscriptions,
+            configureStartParameter(startParameter, taskSpecs, runDefaultTasks),
             ImmutableSet.copyOf(testExecutionRequest.getTestExecutionDescriptors()),
             classNames,
             getInternalJvmTestRequests(testExecutionRequest, classNames),
             getDebugOptions(testExecutionRequest),
             getTaskAndTests(testExecutionRequest),
-            testExecutionRequest.isRunDefaultTasks(false),
-            testExecutionRequest.getTaskSpecs(Collections.emptyList()));
+            runDefaultTasks,
+            taskSpecs);
+    }
+
+    private static StartParameterInternal configureStartParameter(StartParameterInternal startParameter, List<InternalTaskSpec> taskSpecs, boolean runDefaultTasks) {
+        Preconditions.checkArgument(
+            startParameter.getTaskNames().isEmpty(),
+            "Cannot pass task requests with start parameter here, got %s",
+            startParameter.getTaskNames());
+        if (!taskSpecs.isEmpty()) {
+            List<String> taskPaths = taskSpecs.stream().map(InternalTaskSpec::getTaskPath).collect(Collectors.toList());
+
+            Preconditions.checkArgument(
+                !runDefaultTasks,
+                "Cannot run default tasks when task specs %s are provided",
+                taskPaths);
+
+            startParameter.setTaskRequests(Collections.singletonList(new DefaultTaskExecutionRequest(taskPaths)));
+        } else if (runDefaultTasks) {
+            startParameter.setTaskRequests(Collections.singletonList(new RunDefaultTasksExecutionRequest()));
+        } else {
+            startParameter.setTaskRequests(Collections.emptyList());
+        }
+
+        return startParameter;
     }
 
     private static Set<InternalJvmTestRequest> getInternalJvmTestRequests(ProviderInternalTestExecutionRequest testExecutionRequest, Set<String> classNames) {

--- a/subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/TestExecutionRequestActionTest.groovy
+++ b/subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/TestExecutionRequestActionTest.groovy
@@ -24,8 +24,8 @@ import spock.lang.Specification
 
 class TestExecutionRequestActionTest extends Specification {
 
-    StartParameterInternal startParameter = Mock()
-    BuildEventSubscriptions buildClientSubscriptions= Mock()
+    StartParameterInternal startParameter = Stub()
+    BuildEventSubscriptions buildClientSubscriptions = Mock()
     ProviderInternalTestExecutionRequest executionRequest = Mock()
 
     def "maps testClasses to internalJvmTestRequests if empty"(){
@@ -34,6 +34,8 @@ class TestExecutionRequestActionTest extends Specification {
         1 * executionRequest.getInternalJvmTestRequests(_) >> []
         1 * executionRequest.getTaskAndTests(_) >> [:]
         1 * executionRequest.getTestClassNames() >> ["org.acme.Foo"]
+        1 * executionRequest.getTaskSpecs(_) >> []
+        1 * executionRequest.isRunDefaultTasks(_) >> false
 
         when:
         def executionRequestAction = TestExecutionRequestAction.create(buildClientSubscriptions, startParameter, executionRequest);

--- a/subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionRequestActionRunner.java
+++ b/subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionRequestActionRunner.java
@@ -17,19 +17,13 @@
 package org.gradle.tooling.internal.provider.runner;
 
 import org.gradle.api.tasks.testing.TestExecutionException;
-import org.gradle.internal.DefaultTaskExecutionRequest;
-import org.gradle.internal.RunDefaultTasksExecutionRequest;
 import org.gradle.internal.buildtree.BuildActionRunner;
 import org.gradle.internal.buildtree.BuildTreeLifecycleController;
 import org.gradle.internal.invocation.BuildAction;
 import org.gradle.internal.operations.BuildOperationAncestryTracker;
 import org.gradle.internal.operations.BuildOperationListenerManager;
-import org.gradle.tooling.internal.protocol.test.InternalTaskSpec;
 import org.gradle.tooling.internal.protocol.test.InternalTestExecutionException;
 import org.gradle.tooling.internal.provider.action.TestExecutionRequestAction;
-
-import java.util.Collections;
-import java.util.stream.Collectors;
 
 public class TestExecutionRequestActionRunner implements BuildActionRunner {
     private final BuildOperationAncestryTracker ancestryTracker;
@@ -72,15 +66,6 @@ public class TestExecutionRequestActionRunner implements BuildActionRunner {
     }
 
     private void doRun(TestExecutionRequestAction action, BuildTreeLifecycleController buildController) {
-        // TODO - move this to the client
-        if (!action.getTaskSpecs().isEmpty()) {
-            action.getStartParameter().setTaskRequests(Collections.singletonList(new DefaultTaskExecutionRequest(action.getTaskSpecs().stream().map(InternalTaskSpec::getTaskPath).collect(Collectors.toList()))));
-        } else if (action.isRunDefaultTasks()) {
-            action.getStartParameter().setTaskRequests(Collections.singletonList(new RunDefaultTasksExecutionRequest()));
-        } else {
-            action.getStartParameter().setTaskRequests(Collections.emptyList());
-        }
-
         buildController.scheduleAndRunTasks(new TestExecutionBuildConfigurationAction(action));
     }
 

--- a/subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/TestLauncherTestSpecCrossVersionSpec.groovy
+++ b/subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/TestLauncherTestSpecCrossVersionSpec.groovy
@@ -98,4 +98,35 @@ class TestLauncherTestSpecCrossVersionSpec extends TestLauncherSpec {
         assertTestExecuted(className: 'example.MyTest', methodName: 'foo', task: ':secondTest')
         assertTestExecuted(className: 'example.MyTest', methodName: 'foo2', task: ':secondTest')
     }
+
+    @TargetGradleVersion(">=8.4")
+    @Issue("https://github.com/gradle/gradle/issues/26056")
+    def "Can execute same tests within different tests types with configuration cache enabled"() {
+        when:
+        launchTests { TestLauncher launcher ->
+            launcher.withArguments("--configuration-cache")
+            launcher.withTestsFor { TestSpecs specs ->
+                specs.forTaskPath(':test').includeMethod('example.MyTest' , 'foo')
+            }
+        }
+
+        then:
+        events.testClassesAndMethods.size() == 2
+        assertTestExecuted(className: 'example.MyTest', methodName: 'foo', task: ':test')
+
+        when:
+        events.clear()
+        stdout.reset() // we are interested in the output of the second build only
+        launchTests { TestLauncher launcher ->
+            launcher.withArguments("--configuration-cache")
+            launcher.withTestsFor { TestSpecs specs ->
+                specs.forTaskPath(':secondTest').includeMethod("example.MyTest", 'foo')
+            }
+        }
+
+        then:
+        !stdout.toString().contains("Reusing configuration cache.")
+        events.testClassesAndMethods.size() == 2
+        assertTestExecuted(className: 'example.MyTest', methodName: 'foo', task: ':secondTest')
+    }
 }

